File: src/navigation/routing/aStar.ts
import { NavigationGraph } from '../graph/navigationGraph';
import { NavNode, Vector3D } from '../graph/nodes';

export interface AStarResult {
    success: boolean;
    path: NavNode[];
    totalDistance: number;
    nodesExplored: number;
}

export function euclideanHeuristic(p1: Vector3D, p2: Vector3D): number {
    return Math.hypot(p1.x - p2.x, p1.y - p2.y, p1.z - p2.z);
}

/**
 * A* Pathfinding Algorithm over NavigationGraph.
 */
export function findAStarPath(
    graph: NavigationGraph,
    startInput: NavNode | string,
    destInput: NavNode | string
): AStarResult {
    const startNode = typeof startInput === 'string' ? graph.getNode(startInput) : startInput;
    const destNode = typeof destInput === 'string' ? graph.getNode(destInput) : destInput;

    if (!startNode || !destNode) {
        return { success: false, path: [], totalDistance: 0, nodesExplored: 0 };
    }

    if (startNode.id === destNode.id) {
        return { success: true, path: [startNode], totalDistance: 0, nodesExplored: 1 };
    }

    const openSet: Set<string> = new Set([startNode.id]);
    const cameFrom: Map<string, string> = new Map();

    const gScore: Map<string, number> = new Map();
    const hScore: Map<string, number> = new Map();
    const fScore: Map<string, number> = new Map();

    gScore.set(startNode.id, 0);
    const startH = euclideanHeuristic(startNode.position, destNode.position);
    hScore.set(startNode.id, startH);
    fScore.set(startNode.id, startH);

    let nodesExplored = 0;

    while (openSet.size > 0) {
        // Find node in openSet with lowest fScore
        let currentId: string | null = null;
        let lowestF = Infinity;

        for (const nodeId of openSet) {
            const f = fScore.get(nodeId) ?? Infinity;
            if (f < lowestF) {
                lowestF = f;
                currentId = nodeId;
            }
        }

        if (!currentId) break;

        nodesExplored++;

        if (currentId === destNode.id) {
            // Reconstruct path
            const path: NavNode[] = [];
            let curr: string | undefined = currentId;
            while (curr) {
                const node = graph.getNode(curr);
                if (node) path.unshift(node);
                curr = cameFrom.get(curr);
            }

            const totalDistance = gScore.get(destNode.id) ?? 0;
            return {
                success: true,
                path,
                totalDistance,
                nodesExplored,
            };
        }

        openSet.delete(currentId);
        const currentG = gScore.get(currentId) ?? Infinity;

        const neighbors = graph.getNeighbors(currentId);

        for (const neighborInfo of neighbors) {
            const neighbor = neighborInfo.node;
            const tentativeG = currentG + neighborInfo.distance;

            if (tentativeG < (gScore.get(neighbor.id) ?? Infinity)) {
                cameFrom.set(neighbor.id, currentId);
                gScore.set(neighbor.id, tentativeG);
                
                const h = euclideanHeuristic(neighbor.position, destNode.position);
                hScore.set(neighbor.id, h);
                fScore.set(neighbor.id, tentativeG + h);

                if (!openSet.has(neighbor.id)) {
                    openSet.add(neighbor.id);
                }
            }
        }
    }

    return {
        success: false,
        path: [],
        totalDistance: 0,
        nodesExplored,
    };
}

File: lib/navigation_core/a_star.dart
import '../models/navigation_graph.dart';
import '../models/navigation_node.dart';
import '../models/route.dart';

class AStarPathfinder {
  /// Calculate optimal route using A* pathfinding algorithm
  RoutePath? findPath({
    required NavigationGraph graph,
    required String startNodeId,
    required String targetNodeId,
  }) {
    if (graph.nodes.isEmpty) return null;

    final nodeMap = {for (final n in graph.nodes) n.nodeId: n};
    final startNode = nodeMap[startNodeId];
    final targetNode = nodeMap[targetNodeId];

    if (startNode == null || targetNode == null) return null;

    if (startNodeId == targetNodeId) {
      return RoutePath(
        routeId: 'route_${startNodeId}_$targetNodeId',
        startNodeId: startNodeId,
        destinationNodeId: targetNodeId,
        pathNodes: [startNode],
        totalDistance: 0.0,
      );
    }

    // Build adjacency list: nodeId -> list of neighbors with edge weights
    final adj = <String, List<_EdgeInfo>>{};
    for (final node in graph.nodes) {
      adj[node.nodeId] = [];
    }

    for (final edge in graph.edges) {
      if (!edge.accessible) continue;
      final from = edge.fromNodeId;
      final to = edge.toNodeId;
      final dist = edge.distance > 0 ? edge.distance : _heuristic(nodeMap[from], nodeMap[to]);

      if (adj.containsKey(from)) {
        adj[from]!.add(_EdgeInfo(targetId: to, distance: dist));
      }
      if (adj.containsKey(to)) {
        adj[to]!.add(_EdgeInfo(targetId: from, distance: dist));
      }
    }

    final gScore = <String, double>{startNodeId: 0.0};
    final fScore = <String, double>{startNodeId: _heuristic(startNode, targetNode)};
    final cameFrom = <String, String>{};
    final openSet = <String>{startNodeId};

    while (openSet.isNotEmpty) {
      // Find node in openSet with lowest fScore = g(n) + h(n)
      String current = openSet.first;
      double minF = fScore[current] ?? double.infinity;

      for (final id in openSet) {
        final score = fScore[id] ?? double.infinity;
        if (score < minF) {
          minF = score;
          current = id;
        }
      }

      if (current == targetNodeId) {
        // Reconstruct path
        final pathNodes = <NavigationNode>[];
        String? curr = targetNodeId;
        while (curr != null) {
          final n = nodeMap[curr];
          if (n != null) pathNodes.add(n);
          curr = cameFrom[curr];
        }
        final reversedPath = pathNodes.reversed.toList();
        final totalDist = gScore[targetNodeId] ?? 0.0;

        return RoutePath(
          routeId: 'route_${startNodeId}_$targetNodeId',
          startNodeId: startNodeId,
          destinationNodeId: targetNodeId,
          pathNodes: reversedPath,
          totalDistance: totalDist,
        );
      }

      openSet.remove(current);

      final currentG = gScore[current] ?? double.infinity;
      final neighbors = adj[current] ?? [];

      for (final neighbor in neighbors) {
        final neighborId = neighbor.targetId;
        final tentativeG = currentG + neighbor.distance;

        if (tentativeG < (gScore[neighborId] ?? double.infinity)) {
          cameFrom[neighborId] = current;
          gScore[neighborId] = tentativeG;
          final neighborNode = nodeMap[neighborId];
          fScore[neighborId] = tentativeG + _heuristic(neighborNode, targetNode);
          openSet.add(neighborId);
        }
      }
    }

    return null; // No accessible route found
  }

  /// Euclidean distance heuristic h(n)
  static double _heuristic(NavigationNode? a, NavigationNode? b) {
    if (a == null || b == null) return 0.0;
    return a.position.distanceTo(b.position);
  }
}

class _EdgeInfo {
  final String targetId;
  final double distance;
  const _EdgeInfo({required this.targetId, required this.distance});
}
